framework/src/controls/PIDTransientControl.C

```diff
@@ -0,0 +1,148 @@
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+// MOOSE includes
+#include "PIDTransientControl.h"
+#include "Function.h"
+#include "Transient.h"
+
+registerMooseObject("MooseApp", PIDTransientControl);
+
+InputParameters
+PIDTransientControl::validParams()
+{
+  InputParameters params = Control::validParams();
+  params.addClassDescription(
+      "Sets the value of a 'Real' input parameter (or postprocessor) based on a Proportional "
+      "Integral "
+      "Derivative control of a postprocessor to match a target a target value.");
+  params.addRequiredParam<PostprocessorName>(
+      "postprocessor", "The postprocessor to watch for controlling the specified parameter.");
+  params.addRequiredParam<FunctionName>("target",
+                                        "The target value 1D time function for the postprocessor");
+  params.addRequiredParam<Real>("K_integral", "The coefficient multiplying the integral term");
+  params.addRequiredParam<Real>("K_proportional",
+                                "The coefficient multiplying the difference term");
+  params.addRequiredParam<Real>("K_derivative", "The coefficient multiplying the derivative term");
+  params.addParam<std::string>(
+      "parameter",
+      "The input parameter(s) to control. Specify a single parameter name and all "
+      "parameters in all objects matching the name will be updated");
+  params.addParam<std::string>("parameter_pp",
+                               "The postprocessor to control. Should be accessed by reference by "
+                               "the objects depending on its value.");
+  params.addParam<Real>(
+      "start_time", -std::numeric_limits<Real>::max(), "The time to start the PID controller at");
+  params.addParam<Real>(
+      "stop_time", std::numeric_limits<Real>::max(), "The time to stop the PID controller at");
+  params.addParam<bool>(
+      "reset_every_timestep",
+      false,
+      "Reset the PID integral when changing timestep, for coupling iterations within a timestep");
+  params.addParam<bool>("reset_integral_windup",
+                        true,
+                        "Reset the PID integral when the error crosses zero and the integral is "
+                        "larger than the error.");
+
+  return params;
+}
+
+PIDTransientControl::PIDTransientControl(const InputParameters & parameters)
+  : Control(parameters),
+    _current(getPostprocessorValueByName(getParam<PostprocessorName>("postprocessor"))),
+    _target(getFunction("target")),
+    _Kint(getParam<Real>("K_integral")),
+    _Kpro(getParam<Real>("K_proportional")),
+    _Kder(getParam<Real>("K_derivative")),
+    _start_time(getParam<Real>("start_time")),
+    _stop_time(getParam<Real>("stop_time")),
+    _reset_every_timestep(getParam<bool>("reset_every_timestep")),
+    _reset_integral_windup(getParam<bool>("reset_integral_windup"))
+{
+  _integral = 0;
+
+  // To keep the previous values if a multiapp coupling iteration fails
```

Comment:
This comment is confusing. How does this keep any old values? This sets old values to `0`.

And why are not these values set in the initializer list?

Reply:
true this comment fits better in the header